node/lifecycle_test.go

```diff
@@ -0,0 +1,280 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/virtual-kubelet/virtual-kubelet/examples/providers/mock"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	logruslogger "github.com/virtual-kubelet/virtual-kubelet/log/logrus"
+	"gotest.tools/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+	watchutils "k8s.io/client-go/tools/watch"
+	"k8s.io/klog"
+)
+
+var (
+	_ record.EventRecorder = (*fakeDiscardingRecorder)(nil)
+)
+
+const (
+	// There might be a constant we can already leverage here
+	testNamespace        = "default"
+	informerResyncPeriod = time.Duration(1 * time.Second)
+	testNodeName         = "testnode"
+	podSyncWorkers       = 3
+)
+
+func init() {
+	klog.InitFlags(nil)
+}
+
+// fakeDiscardingRecorder discards all events. Silently.
+type fakeDiscardingRecorder struct {
+	logger log.Logger
+}
+
+func (r *fakeDiscardingRecorder) Event(object runtime.Object, eventType, reason, message string) {
+	r.Eventf(object, eventType, reason, message)
+}
+
+func (r *fakeDiscardingRecorder) Eventf(object runtime.Object, eventType, reason, messageFmt string, args ...interface{}) {
+	r.logger.WithFields(map[string]interface{}{
+		"object":    object,
+		"eventType": eventType,
+		"message":   fmt.Sprintf(messageFmt, args...),
+	}).Infof("Received event")
+}
+
+func (r *fakeDiscardingRecorder) PastEventf(object runtime.Object, timestamp metav1.Time, eventType, reason, messageFmt string, args ...interface{}) {
+	r.logger.WithFields(map[string]interface{}{
+		"timestamp": timestamp.String(),
+		"object":    object,
+		"eventType": eventType,
+		"message":   fmt.Sprintf(messageFmt, args...),
+	}).Infof("Received past event")
+}
+
+func (r *fakeDiscardingRecorder) AnnotatedEventf(object runtime.Object, annotations map[string]string, eventType, reason, messageFmt string, args ...interface{}) {
+	r.logger.WithFields(map[string]interface{}{
+		"object":      object,
+		"annotations": annotations,
+		"eventType":   eventType,
+		"message":     fmt.Sprintf(messageFmt, args...),
+	}).Infof("Received annotated event")
+}
+func TestPodLifecycle(t *testing.T) {
+	// We don't do the defer cancel() thing here because t.Run is non-blocking, so the parent context may be cancelled
+	// before the children are finished and there is no way to do a "join" and wait for them without using a waitgroup,
+	// at which point, it doesn't seem much better.
+	ctx := context.Background()
+
+	newLogger := logruslogger.FromLogrus(logrus.NewEntry(logrus.StandardLogger()))
```

Comment:
Could make a logger for `t.Log`.

Reply:
I find it confusing because if you try to compare klog vs. t.log messages they're in a different order.

Reply:
Yep. I remember I tried this now.